### PR TITLE
[PropertyAccess] Fix accessing dynamic properties

### LINKED ISSUE #37026: Recover support to write properties for non stdClass objects
**Symfony version(s) affected**: 5.1 and lower

**Description**  
When generating entities (e.g. via nelmio/alice) it was long possible to write values into existing properties.
But now it seems like the property accessor does only allows this for instances of stdClass .

Possibly since this change: https://github.com/symfony/property-access/commit/0583535d2cfa597ccd6a8e75f6f3bf5a138a5541

So when a class like

```
class Foo {
}
```

has no property but an object gets a property later on like

```
$o = new Foo();
$o->bla = null;
```

then the PropertyAcessor is no longer able to write into this existing property
and instead throws the NoSuchPropertyException exception with a "Could not determine access type" message.

**Possible Solution**  

Drop the explicit check for stdClass 

```
} elseif (property_exists($object, $property)) {
            $object->$property = $value;
        }
```

see https://github.com/symfony/property-access/blob/0583535d2cfa597ccd6a8e75f6f3bf5a138a5541/PropertyAccessor.php#L509

### PR DESCRIPTION
**Pull request by andreyserdjuk**
| Q             | A
| ------------- | ---
| Branch?       | 5.1 (to be switched when merging)
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #37026 
| License       | MIT
| Doc PR        | no

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
/cc @joelwurtz @Korbeil WDYT?

On my side, I'm wondering about the perf impact of the implementation. Isn't this the same as removing the `$object instanceof \stdClass` check?

Please also add tests.

**Comment by andreyserdjuk** [comment]
@nicolas-grekas the only reason why \stdClass check was written before - to be 100% sure that class property is dynamic and cannot be private/protected - inaccessible.
This happens because of `$access = $this->getReadInfo($class, $property);` considers only 
But in case

**Comment by andreyserdjuk** [comment]
@nicolas-grekas it will fail when trying to access protected/private property (`$access` is null but we dunno why: it's a dynamic object property or class protected/private inaccessible property). The only reason why \stdClass check was written before - to be 100% sure that class property is dynamic and cannot be private/protected - inaccessible.
To improve performance I suggest having `PropertyReadInfo` in any case to make a decision looking on `PropertyReadInfo::$visibility`. But it requires some refactoring.

**Comment by andreyserdjuk** [comment]
Update: unit-test added. But only PropertyAccessor::getValue() is covered with getting dynamic value. setValue() - I didn't cover and as I see from unit-tests, the dynamic value cannot be set by design - at least setter method should be.  
@dunglas , could you please tell (you created ReflectionExtractor), whether support of dynamic values getting/setting by ref evolutionally (and intentionally) was removed? I just want to be sure not to harm the idea of this tool.

**Comment by joelwurtz** [comment]
I'm not sure about perf also, it may be slower (but need benchmark for that)

Another possible implementation would be to follow @nicolas-grekas suggestion by only removing the instanceof part and add a try catch when getting the value (to catch error if property is not accessible : protected / private), maybe it would keep perf reasonable and would also allow a better error message (like property exists but is not accessible)

**Comment by andreyserdjuk** [comment]
@nicolas-grekas please review the last approach suggested by @joelwurtz .

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 1]
```diff
@@ -428,17 +428,23 @@ private function readProperty(array $zval, string $property, bool $ignoreInvalid
...
-                $result[self::REF] = &$object->$property;
+        } elseif (property_exists($object, $property)) {
+            try {
+                $result[self::VALUE] = $object->$property;
+                if (isset($zval[self::REF])) {
+                    $result[self::REF] = &$object->$property;
+                }
+            } catch (\Error $e) {
+                if (!$ignoreInvalidProperty) {
+                    throw new NoSuchPropertyException(sprintf('Can\'t read protected or private property "%s" in class "%s".', $property, $class));
```
Comment (joelwurtz) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php:446]: you can add the error in the parent exception

File: src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorTest.php [hunk 2]
```diff
@@ -97,6 +98,18 @@ public function testGetValue($objectOrArray, $path, $value)
...
+     */
+    public function testGetDynamicValue()
+    {
+        $value = 'dynamicPropertyValue';
+        $path = 'dynamicProperty';
+        $object = new TestClassDynamicProperty($value);
+
+        $this->assertSame($value, $this->propertyAccessor->getValue($object, $path));
+    }
+
```
Comment (joelwurtz) [inline src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorTest.php:112]: Maybe you can add a test for the exception path

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 3]
```diff
@@ -428,17 +428,23 @@ private function readProperty(array $zval, string $property, bool $ignoreInvalid
...
-            if (isset($zval[self::REF])) {
-                $result[self::REF] = &$object->$property;
+        } elseif (property_exists($object, $property)) {
+            try {
+                $result[self::VALUE] = $object->$property;
+                if (isset($zval[self::REF])) {
+                    $result[self::REF] = &$object->$property;
+                }
+            } catch (\Error $e) {
+                if (!$ignoreInvalidProperty) {
```
Comment (joelwurtz) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php:445]: maybe it need also a check on the error message (like done upper) wdyt @nicolas-grekas ?

File: src/Symfony/Component/PropertyAccess/composer.json [hunk 4]
```diff
@@ -21,7 +21,8 @@
         "symfony/property-info": "^5.1.1"
     },
     "require-dev": {
-        "symfony/cache": "^4.4|^5.0"
+        "symfony/cache": "^4.4|^5.0",
+        "phpunit/phpunit": "8.5.x-dev"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyAccess/composer.json:25]: please remove this change

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php:445]: sure

File: src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorTest.php (hunk 2 above)
Comment (andreyserdjuk) [reply src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorTest.php:112]: The thrown exception is already covered on access to the existent non-public property (in another unit-test). To assert exact Exception was thrown after attempt property by reference I suggest TestCase::expectExceptionMessage().

File: src/Symfony/Component/PropertyAccess/composer.json (hunk 4 above)
Comment (andreyserdjuk) [reply src/Symfony/Component/PropertyAccess/composer.json]: done

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 3 above)
Comment (andreyserdjuk) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php:445]: done

File: src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorTest.php (hunk 2 above)
Comment (andreyserdjuk) [reply src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorTest.php:112]: done

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 1 above)
Comment (andreyserdjuk) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php:446]: Sorry, didn't get what do you mean?

**Review by maxhelias — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @andreyserdjuk.

**Comment by fabpot** [comment]
Reverting this PR as it breaks things by bypassing a lot of logic. One such example is when you have a class with a private property AND a __get method to access it (like TestClassMagicGet in our test suite). Before this PR, if you disable magic __get method, you will get an expected NoSuchPropertyException. But now, the property_exists() call will return true, and $object->$property will get the value because there is a __get method (which the code explicitly disallow).

**Comment by Voetloos** [comment]
Issue #37615 is again an issue since this is reverted. Any update on that? @fabpot @xabbuh @nicolas-grekas

**Pull request merged**